cpp/src/arrow/compute/kernels/scalar_random.cc

```diff
@@ -66,23 +66,23 @@ Status ExecRandom(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
     return Status::Invalid("Negative number of elements");
   }
 
-  auto out_data = ArrayData::Make(float64(), options.length, 0);
-  out_data->buffers.resize(2, nullptr);
-
-  ARROW_ASSIGN_OR_RAISE(out_data->buffers[1],
-                        ctx->Allocate(options.length * sizeof(double)));
-  double* out_buffer = out_data->template GetMutableValues<double>(1);
-
   if (options.initializer == RandomOptions::Seed) {
     gen.seed(options.seed);
   } else {
     std::lock_guard<std::mutex> seed_gen_lock(seed_gen_mutex);
     gen.seed(seed_gen());
   }
+  // TODO(wesm): refactor to use batch length instead of passing length in
+  // options
+  auto out_data = ArrayData::Make(float64(), options.length, 0);
```

Comment:
Note: I found this aspect of `random()` implementation to be a huge eyesore -- will create a Jira to fix this to write into preallocated memory and not pass the length through the options